src/wutil.cpp

```diff
@@ -711,6 +713,66 @@ double fish_wcstod(const wchar_t *str, wchar_t **endptr) {
+/// which is legal to include in a parse (according to the above rules). Free-floating leading
+/// underscores ("_ 3") are not allowed and will result in a no-parse. Underscores are not allowed
+/// before or inside of "infinity" or "nan" input. Trailing underscores after "infinity" or "nan"
+/// are not consumed.
+double fish_wcstod_underscores(const wchar_t *str, wchar_t **endptr) {
+    const wchar_t *orig = str;
+    // Skip leading whitespace.
+    while (iswspace(*str)) str++;
+    size_t leading_whitespace = (size_t)(str - orig);
+    auto is_sign = [](wchar_t c) { return c == L'+' || c == L'-'; };
+    auto is_inf_or_nan_char = [](wchar_t c) {
+        return c == L'i' || c == L'I' || c == L'n' || c == L'N';
+    };
+    // We don't do any underscore-stripping for infinity/NaN.
+    if (is_inf_or_nan_char(*str) || (is_sign(*str) && is_inf_or_nan_char(*(str + 1)))) {
+        return fish_wcstod(orig, endptr);
+    }
+    std::wstring pruned;
+    std::vector<size_t> underscores;
+    // We keep track of the positions *in the pruned string* where there used to be underscores. We
+    // will pass the pruned version of the input string to the system wcstod, which in turn will
+    // tell us how many characters it consumed. Then we will set our own endptr based on (1) the
+    // number of characters consumed from the pruned string, and (2) how many underscores came
+    // before the last consumed character. The alternative to doing it this way (for example, "only
+    // deleting the correct underscores") would require actually parsing the input string -- for
+    // example, sometimes a dot is a decimal delimiter and sometimes it represents the end of the
+    // valid float, as in "1_2.3_4.5_6".
+    auto take_char = [&pruned, &str, &underscores]() {
+        if (*str == L'_') underscores.push_back(pruned.length());
+        else pruned.push_back(*str);
+        str++;
+    };
+    // Take all alphanumeric characters that can appear in a strtod numeric literal, and +/- signs.
+    auto is_possible_alnum = [](wchar_t c) {
+        // If we wanted to be future-proof against a hypothetical strtod from the future that, say,
+        // allows octal literals using 0o, etc., we could just use iswalnum instead.
+        return iswxdigit(c) || c == L'P' || c == L'p' || c == L'X' || c == L'x';
+    };
+    // In order to be more clever, we would need to do a full parse, because sometimes 'a' is a hex
+    // digit and sometimes it is the end of the parse, sometimes '+' is an exponent sign and
+    // sometimes it is the end of the parse, etc.
+    while (is_possible_alnum(*str) || is_sign(*str) || *str == L'.' || *str == L'_') take_char();
+    const wchar_t *pruned_begin = pruned.c_str();
+    const wchar_t *pruned_end = nullptr;
+    double result = fish_wcstod(pruned_begin, (wchar_t **)(&pruned_end));
+    if (pruned_end == pruned_begin) {
+        *endptr = (wchar_t *)orig;
```

Comment:
shouldn't we check if endptr is null here?
(I guess we could also move this outside the if, and then use += below, not sure)

Reply:
Good catch, sorry about that.

I don't think we should move it because then we need extra logic about not offsetting by the leading whitespace, and the logic of the implementation is easier to read this way.